### ISSUE TITLE
with-lock question or feature request

### ISSUE THREAD

**Issue by Grunthos** [borgbackup/borg#5293]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

QUESTION (or maybe request)

I use the ```with-lock``` feature to execute a script that synchronizes with other borg operations.

The basic approach is:

- start script
- see if ```BORG_VERSION``` is defined (this is set by ```borg with-lock``` in the subshell
- if not defined: use ```borg with-lock``` to restart the script with all of its parameters
- else: do stuff

My concern is that ```BORG_VERSION``` seems to be a side-effect, not a documented result of using ```with-lock```.

Is there a recommended approach to resume-with-sync in a script? And, possibly, could something like ```BORG_LOCKED_REPOS``` be added to the environment variables of subprocesses when any repo is locked? (It may contain more than one)

**Comment by ThomasWaldmann** [comment]
Not sure what you mean by resume-with-sync.

borg operations always have to do with precisely 1 repo, so BORG_LOCKED_REPOS sounds out of scope.

**Comment by Grunthos** [comment]
1. By resume-with-sync (probably a bade name) I mean precisely the process described in my post (restarting a scripts after using with-lock.

2. I could live with ```BORG_LOCKED_REPO``` then, but I thought you might need to deal with the following:
```
borg with-lock REPO1 bash
borg with-lock REPO2 bash
# At this point I woukld suggest BORG_LOCKED_REPO(s) would need to be well-defined and consistent.
```

**Comment by ThomasWaldmann** [comment]
you did notice that the lock is only held as long as the invoked sub-process (here e.g. the first `bash`) is running?

**Comment by Grunthos** [comment]
Yes indeed. It's a trivial example; the second command against REPO2 executes in the bash shell of the first command. In the final shell, two repos will be locked. 

As I explained in the first post, in my case I restart a script using with-lock (passing all the same parameters), so it runs as a sub-process of the fist execution. I use BORG_VERSION as a (very poor) proxy for "is the repo locked". The suggestion of BORG_LOCKED_REPOS is a more robust solution, and my question was "is there another way to check"?

**Comment by ThomasWaldmann** [comment]
OK, guess I can follow now.

`borg with-lock repoN somecommand` would add repoN to the set of repos represented in BORG_LOCKED_REPOS before invoking `somecommand`.

Theoretically it would also remove it from the set again after `somecommand` has returned but guess we are only modifying the env of the `borg with-lock` process, which terminates then anyway (not the env of the shell / process that invoked borg - so it was never added there and thus we do not need to remove it either).

```
BORG_LOCKED_REPOS=repo1\nrepo2\n...repoN  # set of locked repos in no special order
```

**Comment by Grunthos** [comment]
Yes, that would do it. The only worry I can see is if some repo specification somehow had a valid "\n" in it. Not sure that's possib;e though, unless via a comment(?if there are any) or a rsh specification.

**Comment by ThomasWaldmann** [comment]
Well, not sure if \n within a repo spec could be valid, but I'ld still see that as "asking for trouble".

Posix byte paths allow all kinds of sh*t so the only "safe" character might be hex 00 which is definitely invalid within a path because it would terminate the path (implementation detail of using C).

Not sure what's easier to deal with hex 00 or hex 0a (by scripts, by other processes invoking borg).

**Comment by Grunthos** [comment]
0x0a would seem to be the best option; I suspect the environment variable creation should do newline escaping on the repo spec, just in case.

**Comment by ThomasWaldmann** [comment]
Closing this for now.

Consider testing borg2 when it is in late beta and check if the locking works for your use case.
